Allowing different ports to be used in nodes

szczurmys opened the pull request:
It could be useful, if we can use a different port for each node.
So, I modified scripts for this and now you can specify own port using enviroment `REPMGR_PORT_NUMBER` and adding port to `REPMGR_PARTNER_NODES`

E.g.
```
# Node 1
      - REPMGR_PRIMARY_HOST=172.17.0.1
      - REPMGR_PRIMARY_PORT=5432
      - REPMGR_PARTNER_NODES=172.17.0.1:5432,172.17.0.1:5433
      - REPMGR_NODE_NAME=pg-0
      - REPMGR_NODE_NETWORK_NAME=172.17.0.1
      - REPMGR_PORT_NUMBER=5432
# Node 2
      - REPMGR_PRIMARY_HOST=172.17.0.1
      - REPMGR_PRIMARY_PORT=5432
      - REPMGR_PARTNER_NODES=172.17.0.1:5432,172.17.0.1:5433
      - REPMGR_NODE_NAME=pg-1
      - REPMGR_NODE_NETWORK_NAME=172.17.0.1
      - REPMGR_PORT_NUMBER=5433
```

Feel free to comment/modify that PR to fulfill your vision.


I tested this using docker-compose.yml:
(You have to change IP 172.17.0.1 to yours ip (or leave that, because it is docker ip on linux host machine, so it could be the same like yours))

```
version: '2'
services:
  pg-0:
    image: bitnami/postgresql-repmgr:12
    ports:
      - 5432:5432
    volumes:
      - pg_0_data:/bitnami/postgresql
    environment:
      - POSTGRESQL_POSTGRES_PASSWORD=adminpassword
      - POSTGRESQL_USERNAME=customuser
      - POSTGRESQL_PASSWORD=custompassword
      - POSTGRESQL_DATABASE=customdatabase
      - REPMGR_PASSWORD=repmgrpassword
      - REPMGR_PRIMARY_HOST=172.17.0.1
      - REPMGR_PRIMARY_PORT=5432
      - REPMGR_PARTNER_NODES=172.17.0.1:5432,172.17.0.1:5433
      - REPMGR_NODE_NAME=pg-0
      - REPMGR_NODE_NETWORK_NAME=172.17.0.1
      - REPMGR_PORT_NUMBER=5432
      - REPMGR_LOG_LEVEL=DEBUG
  pg-1:
    image: bitnami/postgresql-repmgr:12
    ports:
      - 5433:5432
    volumes:
      - pg_1_data:/bitnami/postgresql
    environment:
      - POSTGRESQL_POSTGRES_PASSWORD=adminpassword
      - POSTGRESQL_USERNAME=customuser
      - POSTGRESQL_PASSWORD=custompassword
      - POSTGRESQL_DATABASE=customdatabase
      - REPMGR_PASSWORD=repmgrpassword
      - REPMGR_PRIMARY_HOST=172.17.0.1
      - REPMGR_PRIMARY_PORT=5432
      - REPMGR_PARTNER_NODES=172.17.0.1:5432,172.17.0.1:5433
      - REPMGR_NODE_NAME=pg-1
      - REPMGR_NODE_NETWORK_NAME=172.17.0.1
      - REPMGR_PORT_NUMBER=5433
      - REPMGR_LOG_LEVEL=DEBUG
volumes:
  pg_0_data:
    driver: local
  pg_1_data:
    driver: local

```


After run the compose file, execute command(s) ;) :
```
echo 'Creating user which is needed by repmgr command.'; \
docker exec -u root -it bitnami-docker-postgresql-repmgr_pg-0_1 bash -c 'useradd -b /opt -m -s /bin/bash repmgr -u 1001 && chown -R repmgr:100 /opt/bitnami/*/tmp' && \
docker exec -u root -it bitnami-docker-postgresql-repmgr_pg-1_1 bash -c 'useradd -b /opt -m -s /bin/bash repmgr -u 1001 && chown -R repmgr:100 /opt/bitnami/*/tmp' && \
\
echo 'Checking status.'; \
docker exec -it bitnami-docker-postgresql-repmgr_pg-0_1 bash -c 'repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf cluster show'; \
docker exec -it bitnami-docker-postgresql-repmgr_pg-1_1 bash -c 'repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf cluster show'; \
\
sleep 5s && \
\
echo 'Stopping server 0.'; \
docker stop bitnami-docker-postgresql-repmgr_pg-0_1 && \
\
sleep 15s && \
\
echo 'Checking status.'; \
docker exec -it bitnami-docker-postgresql-repmgr_pg-1_1 bash -c 'repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf cluster show'; \
\
sleep 5s && \
\
echo 'Starting server 0.'; \
docker start bitnami-docker-postgresql-repmgr_pg-0_1 && \
\
sleep 15s && \
\
echo 'Checking status.'; \
docker exec -it bitnami-docker-postgresql-repmgr_pg-0_1 bash -c 'repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf cluster show'; \
docker exec -it bitnami-docker-postgresql-repmgr_pg-1_1 bash -c 'repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf cluster show'; \
\
sleep 5s && \
\
echo 'Stopping server 1.'; \
docker stop bitnami-docker-postgresql-repmgr_pg-1_1 && \
\
sleep 15s && \
\
echo 'Checking status.'; \
docker exec -it bitnami-docker-postgresql-repmgr_pg-0_1 bash -c 'repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf cluster show'; \
\
sleep 5s && \
\
echo 'Starting server 1.'; \
docker start bitnami-docker-postgresql-repmgr_pg-1_1 && \
\
sleep 15s && \
\
echo 'Checking status.'; \
docker exec -it bitnami-docker-postgresql-repmgr_pg-0_1 bash -c 'repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf cluster show'; \
docker exec -it bitnami-docker-postgresql-repmgr_pg-1_1 bash -c 'repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf cluster show'

```

juan131 commented on 10/debian-10/prebuildfs/libnet.sh:
```diff
@@ -42,3 +42,48 @@ is_hostname_resolved() {
+#########################
+canBeIPv6() {
+	[[ $(echo $1  | grep -o ':' | wc -l) -ge 2 ]]
+}
```
We already have a `validate_ipv4` function at **libvalidations.sh**, see:
- https://github.com/bitnami/bitnami-docker-postgresql-repmgr/blob/master/11/debian-10/prebuildfs/libvalidations.sh#L177

If we're implementing some kind of validations for **IPv6**, we should include them there and following a similar format.

juan131 replied:
Do not use "camelCase" for function names

szczurmys replied:
Ok. I will remove canBeIPv6, because you have right, that I should add support also in other places in code, if I would like to use IPv6.

juan131 commented on 10/debian-10/prebuildfs/libnet.sh:
```diff
@@ -42,3 +42,48 @@ is_hostname_resolved() {
+#   $1 - node_address - String
+#   $2 - default_port - Integer
+# Returns:
+#   Array - (host port)
```
You're actually returning a string. Then, you're transforming in an array doing:
```bash
readarray -t address < <(extractHostAndPort $node $REPMGR_PRIMARY_PORT)
```

juan131 replied:
I would use a different approach. You can implement sth like:
```bash
########################
# Parse URL
# Globals:
#   None
# Arguments:
#   $1 - url - String
#   $2 - field to obtain. Valid options (protocol, hostname, or port) - String
# Returns:
#   String
parse_url() {
  ...
}
```
So you can use, sth like:
```bash
protocol="$(parse_url "$my_url" "protocol")"
port="$(parse_url "$my_url" "port")"
host="$(parse_url "$my_url" "host")"
```

szczurmys replied:
That's a good proposition, I will do that. But you have some suggestions for the other places where I did that (https://github.com/bitnami/bitnami-docker-postgresql-repmgr/pull/11/files#diff-a418dd7dd3d5127e0e68970e7ab0ec7dR197)? 

I can return parameters by globals, but I do not sure that will be a good idea.

juan131 commented on 10/debian-10/docker-compose.yml:
```diff
@@ -13,9 +13,11 @@ services:
       - REPMGR_PRIMARY_HOST=pg-0
-      - REPMGR_PARTNER_NODES=pg-0,pg-1
+      - REPMGR_PRIMARY_PORT=5432
+      - REPMGR_PARTNER_NODES=pg-0,pg-1:5432
```
Shouldn't this be?
```
REPMGR_PARTNER_NODES=pg-0:5432,pg-1:5432
```

szczurmys replied:
Yes and no :) , because doing this I would like to show, that has backward compatibility and it is not necessary to use ports at all.

juan131 replied:
Oh great!

szczurmys commented:
Thanks for response.

In our environment we cannot use docker-compose, kubernetes, etc. (and we cannot change this, for now), so this allows us to run multiple postgres in one host machine.

On the production we have separate servers and it is not a big problem, but that changed is very useful for testing :).

Also in future we want to recognize putting load balancer address in REPMGR_PARTNER_NODES (but we will research, if it is possible at all).

szczurmys commented:
> Hi @szczurmys
> 
> We updated the library libnet.sh (see https://github.com/bitnami/bitnami-docker-postgresql-repmgr/blob/master/12/debian-10/prebuildfs/libnet.sh#L58) including a new function **"parse_uri"**.
> 
> Could you please take a look at it? It can very useful for you.
> ```

Hi @juan131 

Yeah, that's will be very useful (I have already used ?query to share ssh port :) ).

But I had to fixed regexp a little:
- added support for password in authority (username:password@localhost)
- made /path optional

juan131 commented on 10/debian-10/prebuildfs/libnet.sh:
```diff
@@ -62,7 +62,7 @@ parse_uri() {
     # additional sub-expressions to split authority into userinfo, host and port
     # Credits to Patryk Obara (see https://stackoverflow.com/a/45977232/6694969)
-    local -r URI_REGEX='^(([^:/?#]+):)?(//((([^:/?#]+)@)?([^:/?#]+)(:([0-9]+))?))?(/([^?#]*))(\?([^#]*))?(#(.*))?'
+    local -r URI_REGEX='^(([^:/?#]+):)?(//((([^@/?#]+)@)?([^:/?#]+)(:([0-9]+))?))?(/([^?#]*))?(\?([^#]*))?(#(.*))?'
```
This is great!!! Thanks so much for the improvements!
Could you please adapt the command below? They're not aligned after you include the changes.

szczurmys replied:
Sure! I forgot about it.

szczurmys commented on 10/debian-10/rootfs/librepmgr.sh:
```diff
@@ -180,44 +182,53 @@ repmgr_validate() {
+            local host=$(parse_uri "$node" 'host')
+            local port=$(parse_uri "$node" 'port')
+            port=${port:-$REPMGR_PRIMARY_PORT}
+            repmgr_debug "Checking node $host,$port..."
```
Done

juan131 commented:
Hi @szczurmys 

We just released new versions of the PostgreSQL with Repmgr container including your changes! Find them below:

- **9.6.17-debian-10-r15**
- **10.12.0-debian-10-r15`**
- **11.7.0-debian-10-r15**
- **12.2.0-debian-10-r15**

szczurmys commented:
Hi @juan131.

That is great! Thank you :)
